### PR TITLE
RFC: add cache for Pkg.Read.available

### PR DESCRIPTION
**Pull request by KristofferC**
Many of the functions in `Pkg` call out to `Pkg.Read.available()` which searches though all of METADATA and reparses all the files. This PR introduces a cache so that in case of the METADATA being in the same state as the previous call, the previous result is then used.

Unfortunately, for some reason (probably the way we structure METADATA with a huge number of files) just runnning the equivalent of `git status` in libgit2 takes ~ 0.1 - 0.3 seconds depending on computer so  even when using the cache it takes a bit of time. FWIW, the `status` command on https://github.com/KristofferC/METADATA_compressed takes 0.005 seconds.

I have more optimizations for Pkg stuff in the pipeline but let's take them one by one.

### CODE REVIEW

**Comment by KristofferC** [comment]
Some timings:

```jl
julia> cd("/home/kristoffer/.julia/v0.5")

# Compilation times + generating cache
julia> @time Base.Pkg.Read.available();
  2.215042 seconds (2.34 M allocations: 120.878 MB, 13.28% gc time)

# Uses cache
julia> @time Base.Pkg.Read.available();
  0.349956 seconds (7.92 k allocations: 1.160 MB)

# The old version
julia> @time Base.Pkg.Read._available(readdir("METADATA"));
  0.927031 seconds (2.20 M allocations: 114.257 MB, 16.73% gc time)

# Repo is now dirty so ignore cache
shell> touch METADATA/foo

julia> @time Base.Pkg.Read.available();
  1.119367 seconds (2.20 M allocations: 114.261 MB, 13.31% gc time)

# Repo clean again
shell> rm METADATA/foo

julia> @time Base.Pkg.Read.available();
  0.344274 seconds (7.92 k allocations: 1.160 MB)
```

**Comment by KristofferC** [comment]
Failing on windows. Has anyone seen this before? Am I not freeing something correctly?

```
LoadError: unlink: operation not permitted (EPERM)
 in uv_error at .\libuv.jl:68 [inlined]
 in unlink at .\file.jl:382
 in #rm#7 at .\file.jl:107
 in #rm at .\<missing>:0
 in #rm#7 at .\file.jl:117
 in #rm at .\<missing>:0
 in #rm#7 at .\file.jl:117
 in #rm at .\<missing>:0
 in #rm#7 at .\file.jl:117
 in #rm at .\<missing>:0
 in #rm#7 at .\file.jl:117
 in #rm at .\<missing>:0
 in #rm#7 at .\file.jl:117
 in #rm at .\<missing>:0
 in #rm#7 at .\file.jl:117
 in #rm at .\<missing>:0
 in #3 at C:\projects\julia\test\pkg.jl:16
 in withenv at .\env.jl:152
 in temp_pkg_dir at C:\projects\julia\test\pkg.jl:7
 in include_string at .\loading.jl:434
 in include_from_node1 at .\loading.jl:484
 in macro expansion at .\util.jl:226 [inlined]
 in runtests at C:\projects\julia\test\testdefs.jl:7
 in #631 at .\multi.jl:1421
 in run_work_thunk at .\multi.jl:1001
 in macro expansion at .\multi.jl:1421 [inlined]
 in #630 at .\event.jl:68
while loading C:\projects\julia\test\pkg.jl, in expression starting on line 6
```

File: base/pkg/read.jl [hunk 1]
```diff
@@ -10,7 +10,22 @@ readstrip(path...) = strip(readstring(joinpath(path...)))
 url(pkg::AbstractString) = readstrip(Dir.path("METADATA"), pkg, "url")
 sha1(pkg::AbstractString, ver::VersionNumber) = readstrip(Dir.path("METADATA"), pkg, "versions", string(ver), "sha1")
 
-function available(names=readdir("METADATA"))
+type AvailableCache
+    sha::String
+    pkgs::Dict{String, Dict{VersionNumber, Available}}
+end
+
+PKG_AVAILABLE_CACHE = AvailableCache("", Dict{VersionNumber,Available}())
```
Comment (tkelman) [inline base/pkg/read.jl]: the second input here isn't the type `pkgs` is supposed to be

Comment (KristofferC) [inline base/pkg/read.jl]: Saved by https://github.com/JuliaLang/julia/blob/cc30dd151e80494d7b52250ecdc13af917dd711c/base/dict.jl#L427 :P

**Comment by KristofferC** [comment]
This is also very weird...

```jl
julia> @time Pkg.add("Plots")
INFO: Nothing to be done
  1.776663 seconds (8.40 k allocations: 1.183 MB)

julia> @time Pkg.add("Plots")
INFO: Nothing to be done
  1.871670 seconds (8.40 k allocations: 1.183 MB, 1.39% gc time)

julia> Profile.start_timer()
0

julia> @time Pkg.add("Plots")
INFO: Nothing to be done
INFO: METADATA might be out-of-date — you may not have the latest version of Plots
INFO: Use `Pkg.update()` to get the latest versions of your packages
  0.482351 seconds (8.54 k allocations: 1.192 MB)

julia> @time Pkg.add("Plots")
INFO: Nothing to be done
INFO: METADATA might be out-of-date — you may not have the latest version of Plots
INFO: Use `Pkg.update()` to get the latest versions of your packages
  0.478928 seconds (8.42 k allocations: 1.185 MB)
```

**Comment by tkelman** [comment]
you probably do need with statements to avoid this leaving libgit2 memory and file handles open for longer than it should be

**Comment by KristofferC** [comment]
Thanks @tkelman you are right. Updated. I am on another computer now and can not reproduce the weirdness with `@profile`. Let's see what AppVeyor thinks this time.

**Comment by KristofferC** [comment]
Pass on Win now. Cool.

**Comment by tkelman** [comment]
what happens to the GitStatus return if there are local modifications that have not been staged or committed?

**Comment by KristofferC** [comment]
It includes it. The `touch foo` in my second post is a new file that is neither staged or commited and as you can see, the slow fallback is then used.

**Comment by KristofferC** [comment]
I believe `GitStatus` must also be finalized

File: base/pkg/read.jl [hunk 2]
```diff
@@ -57,16 +57,18 @@ function available(cache::AvailableCache = PKG_AVAILABLE_CACHE)
...
-                        cache.pkgs = _available(names)
-                        cache.sha = sha
+                LibGit2.with(LibGit2.GitStatus(repo)) do gitstatus
+                    if length(gitstatus) == 0
+                        # Sha does not match, update the cache with the new pkgs
+                        if cache.sha != sha
+                            cache.pkgs = _available(names)
+                            cache.sha = sha
+                        end
+                        # Copy because some functions that uses this data mutate state, like Pkg.Query.requirements
```
Comment (tkelman) [inline base/pkg/read.jl]: some functions that use

**Comment by tkelman** [comment]
could use some test additions - should be possible to track the state of the cache at a few places in the pkg test

**Comment by KristofferC** [comment]
Ok, I will try to add some tests

File: base/pkg/read.jl [hunk 3]
```diff
@@ -10,7 +10,22 @@ readstrip(path...) = strip(readstring(joinpath(path...)))
 url(pkg::AbstractString) = readstrip(Dir.path("METADATA"), pkg, "url")
 sha1(pkg::AbstractString, ver::VersionNumber) = readstrip(Dir.path("METADATA"), pkg, "versions", string(ver), "sha1")
 
-function available(names=readdir("METADATA"))
```
Comment (tkelman) [inline base/pkg/read.jl]: people might be using this existing method

File: base/pkg/read.jl [hunk 4]
```diff
@@ -28,7 +43,38 @@ function available(names=readdir("METADATA"))
...
+available(pkg::AbstractString) = get(_available([pkg]),pkg,Dict{VersionNumber,Available}())
+
+# Uses cached data if not outdated
+function available(cache::AvailableCache = PKG_AVAILABLE_CACHE)
+    names = readdir("METADATA")
+    # Not a git repo so just bail on using cache
+    if in(".git", names)
+        pkg, usecache = LibGit2.with(LibGit2.GitRepo("METADATA")) do repo
+            LibGit2.with(LibGit2.head(repo)) do head
+                sha = string(Base.LibGit2.Oid(head))
```
Comment (tkelman) [inline base/pkg/read.jl]: worth storing this as bytes instead of a string?

**Comment by KristofferC** [comment]
Mac couldnt find libopenblas.

**Comment by KristofferC** [comment]
Actually, `Query.requirements` does a deepcopy already so the copying before return is not necessary.

**Comment by KristofferC** [comment]
Actually, deepcopying the dict takes like 0.3 seconds and is called on each `Pkg.resolve()`! My hand written copy takes like 0.002 so should probably just swap it out.

**Comment by KristofferC** [comment]
Restarted travis. Err was: https://gist.github.com/KristofferC/bc65d4c8a7aff3a920cd6513cb2b81b9

**Comment by KristofferC** [comment]
Could potentially be backported after getting battle tested on master for a while.

**Comment by tkelman** [comment]
the actual error was earlier, the very end of the backtrace doesn't usually contain much useful information about test failures

**Comment by KristofferC** [comment]
Ok to merge?

File: base/pkg/read.jl [hunk 5]
```diff
@@ -30,6 +37,36 @@ function available(names=readdir("METADATA"))
 end
 available(pkg::AbstractString) = get(available([pkg]),pkg,Dict{VersionNumber,Available}())
 
+# Uses cached data if not outdated
+function available(cache::AvailableCache = PKG_AVAILABLE_CACHE)
+    names = readdir("METADATA")
+    # Not a git repo so just bail on using cache
+    if in(".git", names)
+        pkg, usecache = LibGit2.with(LibGit2.GitRepo("METADATA")) do repo
```
Comment (tkelman) [inline base/pkg/read.jl]: `pkgs` plural?

Comment (KristofferC) [inline base/pkg/read.jl]: Fixed

**Comment by KristofferC** [comment]
Bump.

**Comment by tkelman** [comment]
would recommend not merging until https://github.com/JuliaLang/julia/issues/18486 is debugged

**Comment by KristofferC** [comment]
Makes sense.

**Comment by KristofferC** [comment]
image

**Comment by tkelman** [comment]
Adding the `status` call will make this always slower if you have any uncommitted changes, right? (`git status` in metadata takes 0.75-1.5 seconds for me on Windows, and that's on an SSD.) Should it be possible to opt out of the caching behavior with a global setting that's faster to check than git status?

**Comment by KristofferC** [comment]
> `git status` in metadata takes 0.75-1.5 seconds for me on Windows, and that's on an SSD.

Wow... Really need a new metadata.. How long does `cd(Pkg.dir()); Pkg.Read.available()` take?

But yes, it will be slower but I thought that making it faster for people without uncommitted changes (which should be like 99.9% or users) was worth it. Alternatively to global setting is that we just stop using the cache at all for the rest of the julia session if there is an uncommited change at any point.

**Comment by tkelman** [comment]
session based would need to be slow once each time. the opt out doesn't have to be highly visible, just something that could be added to juliarc if you expect to be in this situation (should be uncommon, but I occasionally leave patch files etc around there)

**Comment by KristofferC** [comment]
Added an ENV var to disable. Could you check how long `Pkg.Read.available()` takes? Would be stupid to have this cache if `git status` takes longer than the read on windows machines..

**Comment by tkelman** [comment]
Ah right, been a little busy, haven't had a consistent windows build for very long lately. On master or this branch? release-0.5 should be more or less identical to master for that, right?

**Comment by KristofferC** [comment]
Should be the same. Just want to see the ratio between a git status and a full pkg read.

**Comment by tkelman** [comment]
```
$ usr/bin/julia -e 'cd(Pkg.dir()); @time Pkg.Read.available(); @time Pkg.Read.available(); @time Pkg.Read.available()'
  2.030962 seconds (3.64 M allocations: 184.305 MB, 13.88% gc time)
  1.647395 seconds (3.52 M allocations: 179.441 MB, 6.53% gc time)
  1.627700 seconds (3.52 M allocations: 179.441 MB, 7.50% gc time)

Tony@LAPTOP-O230JCFF ~/julia
$ usr/bin/julia -e 'cd(Pkg.dir("METADATA")); @time run(`git status`); @time run(`git status`); @time run(`git status`)'
On branch metadata-v2
Your branch is up-to-date with 'origin/metadata-v2'.
nothing to commit, working directory clean
  2.079213 seconds (4.64 k allocations: 208.900 KB)
On branch metadata-v2
Your branch is up-to-date with 'origin/metadata-v2'.
nothing to commit, working directory clean
  2.035907 seconds (33 allocations: 1.453 KB)
On branch metadata-v2
Your branch is up-to-date with 'origin/metadata-v2'.
nothing to commit, working directory clean
  2.001022 seconds (33 allocations: 1.453 KB)
```

**Comment by KristofferC** [comment]
Well this PR makes no sense if we want to support arbitrary changes to METADATA files but checking if something has changed is more expensive than departing the departing everything. I had very different numbers on my Linux machine though.

**Comment by tkelman** [comment]
Should we make the caching unix-only then?

**Comment by KristofferC** [comment]
Is the libgit version any faster for its git status?

**Comment by tkelman** [comment]
How many of these `with`s do I need? Those are for avoiding leaking resources, right? Looks even slower!
```
$ usr/bin/julia -e 'repo=LibGit2.GitRepo(Pkg.dir("METADATA")); @time LibGit2.GitStatus(repo); @time LibGit2.GitStatus(repo); @time LibGit2.GitStatus(repo)'
 10.643553 seconds (22.96 k allocations: 1.025 MB)
  6.529404 seconds (8 allocations: 288 bytes)
  6.541869 seconds (8 allocations: 288 bytes)
```

**Comment by KristofferC** [comment]
Just... what? That takes 0.1s on my machine.

**Comment by KristofferC** [comment]
Out of curiosity,  if you have time,  how much does the status take on my experimental metadata-v3 branch..

**Comment by tkelman** [comment]
libgit2 still slower than command-line git, but much faster than metadata-v2
```
$ ~/julia/usr/bin/julia -e 'repo=LibGit2.GitRepo(Pkg.dir("METADATA")); @time LibGit2.GitStatus(repo); @time LibGit2.GitStatus(repo); @time LibGit2.GitStatus(repo)'
  0.434575 seconds (22.96 k allocations: 1.025 MB)
  0.444559 seconds (8 allocations: 288 bytes)
  0.442820 seconds (8 allocations: 288 bytes)

Tony@LAPTOP-O230JCFF ~/.julia/v0.5/METADATA
$ ~/julia/usr/bin/julia -e 'cd(Pkg.dir("METADATA")); @time run(`git status`); @time run(`git status`); @time run(`git status`)'
On branch metadata-v3
Your branch is up-to-date with 'KristofferC/metadata-v3'.
nothing to commit, working directory clean
  0.135355 seconds (4.64 k allocations: 208.900 KB)
On branch metadata-v3
Your branch is up-to-date with 'KristofferC/metadata-v3'.
nothing to commit, working directory clean
  0.050099 seconds (33 allocations: 1.453 KB)
On branch metadata-v3
Your branch is up-to-date with 'KristofferC/metadata-v3'.
nothing to commit, working directory clean
  0.062526 seconds (33 allocations: 1.453 KB)
```

**Comment by KristofferC** [comment]
My results on the windows machine (on metadata-v2):

```jl

julia> repo=LibGit2.GitRepo(Pkg.dir("METADATA")); @time LibGit2.GitStatus(repo); @time LibGit2.GitStatus(repo); @time LibGit2.GitStatus(repo)
  3.521479 seconds (22.96 k allocations: 1008.483 KB)
  1.537661 seconds (8 allocations: 288 bytes)
  1.543216 seconds (8 allocations: 288 bytes)
Base.LibGit2.GitStatus(Ptr{Void} @0x0000000019bb1d40)

julia> cd(Pkg.dir()); @time Pkg.Read.available(); @time Pkg.Read.available(); @time Pkg.Read.available()
 16.954721 seconds (3.36 M allocations: 170.045 MB, 1.23% gc time)
  1.638704 seconds (3.24 M allocations: 165.239 MB, 5.97% gc time)
  1.632870 seconds (3.24 M allocations: 165.239 MB, 5.73% gc time)

julia> cd(Pkg.dir("METADATA")); @time run(`git status`); @time run(`git status`); @time run(`git status`)
  0.157275 seconds (33 allocations: 1.453 KB)
  0.155995 seconds (33 allocations: 1.453 KB)
  0.155281 seconds (33 allocations: 1.453 KB)
```

So for me, `git status` is like 13 times faster than yours.. wut. Secondly,  my first run of `Pkg.Read` took 17 seconds... also wut. Third is that LibGit2.GitStatus also is almost as slow as rereading everything.

**Comment by KristofferC** [comment]
But on MSYS:

```
julia> cd(Pkg.dir("METADATA")); @time run(`git status`); @time run(`git status`); @time run(`git status`)
  4.380275 seconds (33 allocations: 1.453 KB)
  1.677471 seconds (33 allocations: 1.453 KB)
  1.667801 seconds (33 allocations: 1.453 KB)
```

**Comment by tkelman** [comment]
The msys2 / cygwin build of git is a bit slower than the native windows one (we bundle the latter in 0.4). Where did you get the build of git from that was only taking 0.15-0.20s on Windows? Sure you weren't on your metadata-v3 branch there?

**Comment by KristofferC** [comment]
```
julia> repo=LibGit2.GitRepo(Pkg.dir("METADATA"))
Base.LibGit2.GitRepo(Ptr{Void} @0x00000000082cba60)

julia> LibGit2.branch(repo)
"metadata-v2"

julia> @time run(`git status`)
On branch metadata-v2
Your branch is behind 'origin/metadata-v2' by 25 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
nothing to commit, working directory clean
  0.280314 seconds (33 allocations: 1.453 KB)
```

¯\_(ツ)_/¯

**Comment by KristofferC** [comment]
```jl
julia> pwd()
"C:\\Users\\kcarl\\.julia\\v0.5\\METADATA"

julia> run(`git branch`)
* metadata-v2

julia> @time run(`git status`)
On branch metadata-v2
Your branch is behind 'origin/metadata-v2' by 25 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
nothing to commit, working directory clean
  0.254716 seconds (33 allocations: 1.453 KB)
```

¯\_(ツ)_/¯

**Comment by KristofferC** [comment]
The current structure of METADATA together with needing to check the dirty state and the different performance of git / libgit2 on different harddrives / OS makes the potential performance gains of this PR too hard to predict. Would be good to revisit if / when METADATA get restructured to not be basically a worst case scenario for git.

**Pull request closed without merging**